Libs/MRML/Core/vtkMRMLStorageNode.h

```diff
@@ -26,6 +27,71 @@ class vtkStringArray;
 
 // STD includes
 #include <vector>
+#include <algorithm>
+
+///
+/// A class for recording warnings and errors associated with this
+/// vtkMRMLStorableNode.  This cannot be hidden within the
+/// vtkMRMLStorableNode class scope because the Python wrapping code
+/// vtkMRMLStorageNodePython.cxx cannot wrap the GetUserMessages
+/// method.
+class vtkMessages
```

Comment:
I'm still not a fan of this class 😛

Reply:
I hear you.  Let's figure this out.

I fear that `std::map<vtkStringArray>` may suffer from some of the same problems as `std::vector<vtkMessages::Message>`.  Either way, the use of the standard library means that we deal with the system-specific integer types such as `size_type` that the standard library employs.

The `std::vector<vtkMessages::Message>` approach allows us to add additional fields to the `Message` struct (such as a time stamp, \_\_LINE\_\_, or cetera) should we decide that we want them later.

Perhaps a hybrid approach that uses `vtkVariantArray<vtkMessage>` for some new `vtkMessage` struct (not plural!) is the way to go.

Regardless of the previous paragraphs, we can have any new structures inherit from `vtkObject`.

Reply:
It's very tempting to migrate everything to std container, but the vtk ones do a nice job of smoothing out any of the cross-platform issues (the worst of those days are gone, but they still crop up sometimes and it's really nice when all the code gets fixed in one place with some cmake or other magic).  So yes, I like your hybrid approach.

We increasingly rely on having scripted control over the application, so making sure the API is wrapped completely is a nice byproduct of using vtkObject derivations unless there's a strong performance or other reason not to.

Reply:
I  want to stick to the current implementation with `std::vector<vtkMessage>` rather than switching to an implementation with `vtkVariantArray`. 

Details: `vtkVariantArray` is not a C++ template, but a container that handles a mixture of types.  I think that you are suggesting that we make use of its ability to contain ordinary pointers to `vtkObjectBase` objects.  So, `vtkMRMLStorageNode` would implement methods to handle memory management, and the setting and parsing of the components of a `vtkMessage` (singular) subclass of `vtkObject`.  In anticipation of having messages for other than `vtkMRMLStorageNode` objects, I am thinking that we would create a `vtkMessages` (plural) class to handle that memory management, setting, and parsing, and we would have `vtkMRMLStorageNode` use that.

However that reasoning brings us back almost full circle to the current design, which differs only in that it uses a `std::vector<vtkMessage>` struct instead of `vtkVariantArray(vtkObjectBase *)` calls.  I note that the `vtkMRMLStorageNode` already has 3 members that are `std::vector<...>` and 2 methods that return `std::vector<...>`, so I think this additional use of `std::vector<...>` is likely to work.  Because `std::vector` accepts actual objects, not merely pointers to them, it is going to save me having to worry about memory management.  And it has the advantage that it is already written and has been through a bit of debugging.

In short, I want to stick to the current implementation with `std::vector<vtkMessage>` rather than switching to an implementation with `vtkVariantArray`.

Reply:
I don't think vtkVariantArray is applicable here.

VTK style would be to create a vtkMessage class (based on vtkObject, therefore having Python wrapping) and for convenience, add a vtkMessageCollection class (based on vtkCollection, with some additional methods that are useful for setting/getting/procesing messages). However, this would be an overkill for a trivially small message class (that only contains a string and an int).

Probably we could just add a message struct, std::deque<message>, and few methods directly to the storage node, but since you already implemented a separate class, we can just use that (it is somewhat nicer in a separate class, so that it can be moved around easier in future code refactorings). Since vtkMessages is not a VTK class, its name must not begin with "vtk" and it has to be a private implementation (not in the public header file). Its features can be exposed via a few new methods in vtkMRMLStorageNode.

Reply:
Stepping back a bit, I think my main point is that this all feels like too much code to address the actual use case.

I think we could accomplish the task at hand with a simple string to capture any messages.  If there's more than one message they can be separated by newlines.  And each line can be prefixed by 'Warning' or 'Error' and the string can be placed directly in the gui.

I see that the original mockup from @lassoan included the count of the number of errors and warnings but I'm not sure that's very meaningful to users.  The main thing is to flag them when something happens and then they need to read the details of the message.

Reply:
I have spent the morning adding `vtkObject` support to `vtkMessages` and in response to @lassoan's comment just above I am changing the name of latter to `vtkMessageCollection`.  It is now in the pull request as commit https://github.com/Slicer/Slicer/pull/5045/commits/aa236538876d8a31ac7f0fa4053ed000bf4b33a3.

Reply:
It would be useful to be able to quickly distinguish between warning (save was successful but there may be some limitations) and error messages (save failed). Displaying the number of messages can be informative, too, but it is not essential - I agree that we could just display the error or warning icon, without any text).

I feel, too, that this is a bit too much infrastructure for implement this little feature. However, if we consider generalizing and expanding this in the future and use it in other places (for example using it for collecting messages while running CLI modules or other processing), or it is planned to be moved to VTK proper, then it might be justified.

> I am changing the name of latter to vtkMessageCollection

If we agree to keep this class then it should be renamed to something more specific. 1. It is in MRML library, so it should be vtkMRML... (if we plan to move it to VTK then we don't need the MRML prefix but then it would need to be moved to vtkAddon, but for this it would be nice to ask VTK core people if they were interested in adding vtkLogMessage and vtkLogMessageCollection classes to VTK). 2. Since there are many kinds of messages, a more specific name would be better something like vtkMRMLLogMessageCollection.

In the long term, the best would be to add a vtkMessage class to VTK, which could store message information (text, type, group name, class name, file name, line number, etc.) and VTK logging macros would use them as event data in warning/error events. This way VTK would not need to create a formatted string from all the input information and applications would not need to parse this string to extract relevant information (making logging and interpretation of logs simpler and faster).

Reply:
I agree it would be a valuable and easy fix to VTK to add a log management feature like this - all the log messages go through the macros anyway, so you could also keep information about the vtkObject that generated the log message.  In this case it would make sense to integrate with the sink code to intercept output from helper libraries too.  It would be cool also to use that for CLIs but also modules could use that mechanism to create logs associated with workflows in a structured way.  Obviously that's a more ambitious project than Lee signed up for.